lib/mayaUsd/ufe/UsdAttributes.cpp

```diff
@@ -341,5 +365,149 @@ UsdAttributes::getUfeTypeForAttribute(const PXR_NS::UsdAttribute& usdAttr) const
+                    } else if (splitName.back() == PXR_NS::UsdShadeTokens->displacement) {
+                        matPrim.CreateDisplacementOutput(PXR_NS::TfToken(splitName[1]));
+                    } else if (splitName.back() == PXR_NS::UsdShadeTokens->volume) {
+                        matPrim.CreateVolumeOutput(PXR_NS::TfToken(splitName[1]));
+                    }
+                }
+            }
+            // Fallback to creating a nodegraph output.
+            connectApi.CreateOutput(baseNameAndType.first, ufeTypeToUsd(type));
+        } else if (baseNameAndType.second == PXR_NS::UsdShadeAttributeType::Input) {
+            connectApi.CreateInput(baseNameAndType.first, ufeTypeToUsd(type));
+        }
+    }
+
+    // Fallback to creating a custom attribute.
+    prim.CreateAttribute(nameAsToken, ufeTypeToUsd(type));
+
+    return UsdAttributes(item).attribute(name);
+}
+bool UsdAttributes::canRemoveAttribute(const UsdSceneItem::Ptr& item, const std::string& name)
+{
+    if (!item || !item->prim().IsActive() || !UsdAttributes(item).hasAttribute(name)) {
+        return false;
+    }
+
+    PXR_NS::TfToken nameAsToken(name);
+    auto            prim = item->prim();
+    auto            attribute = prim.GetAttribute(nameAsToken);
+    if (attribute.IsCustom()) {
+        // Custom attributes can be removed.
+        return true;
+    }
+
+    // We can also remove NodeGraph boundary attributes
+    PXR_NS::UsdShadeNodeGraph      ngPrim(prim);
+    PXR_NS::UsdShadeConnectableAPI connectApi(prim);
+    if (ngPrim && connectApi) {
+        auto baseNameAndType = PXR_NS::UsdShadeUtils::GetBaseNameAndType(nameAsToken);
+        if (baseNameAndType.second == PXR_NS::UsdShadeAttributeType::Output) {
+            PXR_NS::UsdShadeMaterial matPrim(prim);
+            if (matPrim) {
+                // Can not remove the 3 main material outputs as they are part of the schema
+                if (baseNameAndType.first == PXR_NS::UsdShadeTokens->surface
+                    || baseNameAndType.first == PXR_NS::UsdShadeTokens->displacement
+                    || baseNameAndType.first == PXR_NS::UsdShadeTokens->volume) {
+                    return false;
+                }
+            } else {
+                for (auto&& authoredOutput : connectApi.GetOutputs(true)) {
+                    if (authoredOutput.GetFullName() == name) {
+                        return true;
+                    }
+                }
+            }
+        } else if (baseNameAndType.second == PXR_NS::UsdShadeAttributeType::Input) {
+            for (auto&& authoredInput : connectApi.GetInputs(true)) {
+                if (authoredInput.GetFullName() == name) {
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
+
+bool UsdAttributes::doRemoveAttribute(const UsdSceneItem::Ptr& item, const std::string& name)
+{
+    if (!item || !item->prim().IsActive() || !UsdAttributes(item).hasAttribute(name)) {
```

Comment:
Same comment as before, why not use canRemoveAttribute to avoid code duplication?